src/base/UniswapV4.sol

```diff
@@ -0,0 +1,571 @@
+     * @param position A struct with position and pool related variables.
+     * @param amount0Desired The desired amount of token0 to add as liquidity.
+     * @param amount1Desired The desired amount of token1 to add as liquidity.
+     */
+    function _increaseLiquidity(
+        uint256[] memory balances,
+        address,
+        PositionState memory position,
+        uint256 amount0Desired,
+        uint256 amount1Desired
+    ) internal virtual override {
+        // Check if token0 is native ETH.
+        bool isNative = position.tokens[0] == address(0);
+
+        // Handle approvals.
+        if (!isNative) _checkAndApprovePermit2(position.tokens[0]);
+        _checkAndApprovePermit2(position.tokens[1]);
+
+        // Calculate liquidity to be added.
+        position.liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            uint160(position.sqrtPrice),
+            TickMath.getSqrtPriceAtTick(position.tickLower),
+            TickMath.getSqrtPriceAtTick(position.tickUpper),
+            amount0Desired,
+            amount1Desired
+        );
+
+        // Cache the currencies.
+        Currency currency0 = Currency.wrap(position.tokens[0]);
+        Currency currency1 = Currency.wrap(position.tokens[1]);
+
+        // Generate calldata to mint new position.
+        bytes memory actions = new bytes(3);
+        actions[0] = bytes1(uint8(Actions.INCREASE_LIQUIDITY));
+        actions[1] = bytes1(uint8(Actions.SETTLE_PAIR));
+        actions[2] = bytes1(uint8(Actions.SWEEP));
+        bytes[] memory params = new bytes[](3);
+        params[0] = abi.encode(position.id, position.liquidity, type(uint128).max, type(uint128).max, "");
+        params[1] = abi.encode(currency0, currency1);
+        params[2] = abi.encode(currency0, address(this));
+
+        // Mint the new position.
+        uint256 ethValue = isNative ? amount0Desired : 0;
+        bytes memory increaseLiquidityParams = abi.encode(actions, params);
+        POSITION_MANAGER.modifyLiquidities{ value: ethValue }(increaseLiquidityParams, block.timestamp);
+
+        // Update the balances, token0 might be native ETH.
+        balances[0] = currency0.balanceOfSelf();
+        balances[1] = ERC20(position.tokens[1]).balanceOf(address(this));
+    }
+
+    /* ///////////////////////////////////////////////////////////////
+                          STAKING LOGIC
+    /////////////////////////////////////////////////////////////// */
+
+    /**
+     * @notice Stakes a Liquidity Position.
+     * @param balances The balances of the underlying tokens.
+     * param positionManager The contract address of the Position Manager.
+     * @param position A struct with position and pool related variables.
+     */
+    function _stake(uint256[] memory balances, address, PositionState memory position) internal virtual override {
+        // If token0 is in native ETH, wrap it.
+        if (position.tokens[0] == address(0)) {
```

Comment:
more of a thought: need to double check this, since sometimes it's immediately saved as tokens[0] == weth, but here it only checks for address(0)

Reply:
We only save it as "tokens[0] == weth" on the _stake(), since after that we have to deposit the asset back into the account.